src/sentry/utils/types.py

```diff
@@ -181,7 +181,7 @@ def convert(self, value):
 }
 
 
-def type_from_value(value: typing.Any) -> typing.Callable[[typing.Any], typing.Any]:
+def type_from_value(value: typing.Any) -> Type:
```

Comment:
the typing system could understand this completely if you add some overrides:

```suggestion
@typing.overload
def type_from_value(value: bytes) -> StringType:
    ...


@typing.overload
def type_from_value(value: str) -> StringType:
    ...


@typing.overload
def type_from_value(value: list[T]) -> SequenceType[T]:
    ...


def type_from_value(value: typing.Any) -> Type:
    """Fetch Type based on a primitive value"""
    return _type_mapping[type(value)]


reveal_type(type_from_value([3]))  # revealed type is demo.SequenceType[int]
```

Without the overrides it gets `demo.Type[Any]`, instead.

Reply:
Thanks for the suggestion on adding overloads for type_from_value(). I've added those now, with the exception of `type_from_value(value: bool) -> BoolType` as mypy thinks it overlaps with the `int` overload 🙃

Reply:
leave a comment? it may be something we can improve later.